String Literal types on es6 interfaces for [Symbol.toStringTag]?

**TypeScript Version:**  2.5.3

**Code**

```ts
class MySpecialMap implements Map<Object, Object> {
   private _map;
   readonly [Symbol.toStringTag] = "MySpecialmap";
   constructor() { this._map = new Map(); }
   // declare all other methods of Map methods with some unique behaviors that wrap calls to this._map. 
}
```

**Expected behavior:**

Classes implementing an ES6 interface can declare a distinct value for `Symbol.toStringTag`.

The goal of this property is to provide a distinct value, using a string literal type for the interface is overly specific.

**Actual behavior:**

```
src/MySpecialMap.ts(1,7): error TS2420: Class 'MySpecialMap' incorrectly implements interface 'Map<Object, Object>'.
  Types of property '[Symbol.toStringTag]' are incompatible.
    Type '"MySpecialMap"' is not assignable to type '"Map"'.
```

@aluanhaddad yeah, I understand those things. sorry, if I wasn't clear or didn't use the correct nomenclature. I wanted to use the type system to ensure that the class I built implemented the `Map` interface. I didn't want to inherit the methods of that class -- I wanted to provide an implementation for them and I wanted to let the interface of that class guide that implementation and ensure compatibility if the type ever changes.

I'm not sure what you mean by "does indeed require that value for the tag" because this is typescript type, not something enforced by es6 and it would be incorrect for my class to declare that it is actually a`"Map"` when it's not according to the semantics of `Symbol.toStringTag`.

I tried to create an interface that extends map and provides a different value for `toStringTag` but this doesn't work.

```ts
interface MySpecialMap<Object, Object> extends Map<Object, Object> {
  readonly [Symbol.toStringTag] = "MySpecialMap";
}
```

yielding the same error:

```
src/MySpecialMap.ts(1,11): error TS2430: Interface 'MySpecialMap<Object>' incorrectly extends interface 'Map<Object, Object>'.
  Types of property '[Symbol.toStringTag]' are incompatible.
    Type '"MySpecialMap"' is not assignable to type '"Map"'.
```

If I declare that it's my class that extends `Map` instead I get different errors:

```
src/MySpecialMap.ts(5,7): error TS2415: Class 'MySpecialMap' incorrectly extends base class 'Map<Object, Object>'.
  Types of property '[Symbol.toStringTag]' are incompatible.
    Type '"MySpecialMap"' is not assignable to type '"Map"'.
src/MySpecialMap.ts(5,7): error TS2417: Class static side 'typeof MySpecialMap' incorrectly extends base class static side '{ readonly prototype: Map<any, any>; }'.
  Types of property 'prototype' are incompatible.
    Type 'MySpecialMap' is not assignable to type 'Map<any, any>'.
      Types of property '[Symbol.toStringTag]' are incompatible.
        Type '"MySpecialMap"' is not assignable to type '"Map"'.
```

I don't understand what benefit TypeScript is imbuing to the javascript developer by using string literal types for `toStringTag` values in its interface definitions of es6 classes. `toStringTag` shouldn't be used for case statements or instance checking. The goal is to make debugging easier, right? How does limiting the possible values of this property help debugging?

I think `toStringTag` should be of type `string` for all es6 interfaces so that developers can override it correctly.

(note: I am using `--target es2015`)

--- a/src/lib/es2015.symbol.wellknown.d.ts
+++ b/src/lib/es2015.symbol.wellknown.d.ts
@@ -63,7 +63,7 @@ interface SymbolConstructor {
 }
 
 interface Symbol {
-    readonly [Symbol.toStringTag]: "Symbol";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface Array<T> {
@@ -107,23 +107,23 @@ interface Date {
 }
 
 interface Map<K, V> {
-    readonly [Symbol.toStringTag]: "Map";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface WeakMap<K extends object, V> {
-    readonly [Symbol.toStringTag]: "WeakMap";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface Set<T> {
-    readonly [Symbol.toStringTag]: "Set";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface WeakSet<T extends object> {
-    readonly [Symbol.toStringTag]: "WeakSet";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface JSON {
-    readonly [Symbol.toStringTag]: "JSON";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface Function {
@@ -138,15 +138,15 @@ interface Function {
 }
 
 interface GeneratorFunction {
-    readonly [Symbol.toStringTag]: "GeneratorFunction";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface Math {
-    readonly [Symbol.toStringTag]: "Math";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface Promise<T> {
-    readonly [Symbol.toStringTag]: "Promise";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface PromiseConstructor {
@@ -241,11 +241,11 @@ interface String {
 }
 
 interface ArrayBuffer {
-    readonly [Symbol.toStringTag]: "ArrayBuffer";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface DataView {
-    readonly [Symbol.toStringTag]: "DataView";
+    readonly [Symbol.toStringTag]: string;
 }
 
 interface Int8Array {

